Refactors the Light Eater and gives it some flavortext. Some minor behavior changes included.

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Converts the effects of the nightmares light eater into a component and a couple elements
- Adds some flavor text to the light eater
- Makes the effects of the nightmare's light eater dispel if the armblade is destroyed

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Probably a net increase in code quality
- More flavortext
- The nightmare's light eater effect can be dispelled if you are willing to sacrifice the armblade

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
refactor: The light eater is now a component
balance: The light eater's esoteric effects can be dispelled if you sacrifice the armblade
add: The light eater has some additional flavortext
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Elementized Light Eater

- Makes the effects of the light eater into an element

TemporalOroboros committed:
Componentizes the light eater

- Adds a temporary version of the light eater element that dispels if the source is destroyed
- Changes the light eater armblade to use the component instead of the element

TemporalOroboros committed:
Fix miscellaneous errors

- Added the light eater component filepath to the .dme
- Fixed some incorrect proc paths

TemporalOroboros committed:
Fixes more code errors.

- It's kind of sad that it is legitimately faster to use the github checks to find out if this parses than to compile this on my own machine

TemporalOroboros committed:
Cleans up unused variables

- Removes a couple variables that wound up being unused

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Re-move nightmare code to antagonist file

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into upstream/master

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Move nightmare species into the antagonist file _again_

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

optimumtact commented:
@tgstation/commit-access review and merge please, it's been sitting here for no reason

Rohesie commented on code/datums/components/light_eater.dm:
```diff
@@ -0,0 +1,198 @@
+	if(!light_eater_devour(morsel, eater))
+		return FALSE
+
+	morsel.visible_message(
+		"<span class='danger'>Something dark and hungry swarms out of \the [eater] and over \the [morsel]!</span>",
+		"<span class='userdanger'>Something dark and hungry swarms out of \the [eater] and burrows into you!</span>",
+		"<span class='danger'>You can feel a dark hum gnaw at your sight.</span>"
+	)
```
This might get spammy. You might want to compile this as a list and return it for feedback, perhaps, so a single visible message is sent.

TemporalOroboros replied:
Done

Rohesie commented on code/datums/components/light_eater.dm:
```diff
@@ -0,0 +1,198 @@
+	/// Tracks things this light eater has eaten
+	var/list/eaten_lights
+
+/datum/component/light_eater/Initialize(list/_eaten)
+	if(!isatom(parent) && !istype(parent, /datum/reagent))
+		return COMPONENT_INCOMPATIBLE
+
+	if(length(_eaten))
+		eaten_lights = list()
+		var/list/cached_eaten_lights = eaten_lights
+		for(var/food in _eaten)
+			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
+			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
+			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
+			cached_eaten_lights += food
+	return ..()
```
```suggestion
/datum/component/light_eater/Initialize(list/_eaten)
	. = ..()
	if(!isatom(parent) && !istype(parent, /datum/reagent))
		return COMPONENT_INCOMPATIBLE

	if(length(_eaten))
		eaten_lights = list()
		var/list/cached_eaten_lights = eaten_lights
		for(var/food in _eaten)
			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
			cached_eaten_lights += food

```
Call parent first on initialization whenever you can.

TemporalOroboros replied:
Done

Rohesie commented on code/__HELPERS/light_eater.dm:
```diff
@@ -0,0 +1,28 @@
+// Procs that are re-used by both the light eater component and element to prevent copypasta and duplicate signals
+
+/// Abstracted as a global proc so I can avoid sending the signal from two places and reduce copypasta
+/proc/light_eater_table_buffet(atom/commisary, datum/light_eater)
```
Could you explain a bit why this proc is not atom-typed, and why you are passing the `light_eaten` (component I assume) as an argument?

TemporalOroboros replied:
That would require remembering what I was thinking a month ago.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Refactors the light eater to reduce spam

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

Rohesie commented on code/datums/components/overlay_lighting.dm:
```diff
@@ -135,6 +136,7 @@
 		COMSIG_ATOM_SET_LIGHT_ON,
 		COMSIG_ATOM_SET_LIGHT_FLAGS,
 		COMSIG_ATOM_USED_IN_CRAFT,
+		COMSIG_LIGHT_EATER_ACT,
```
Are you are registering to `COMSIG_LIGHT_EATER_QUEUE` and unregistering to `COMSIG_LIGHT_EATER_ACT`
This does not seem intended

TemporalOroboros replied:
Fixed.

Rohesie commented on code/datums/elements/light_eater.dm:
```diff
@@ -0,0 +1,171 @@
+ */
+/datum/element/light_eater/proc/table_buffet(atom/commisary, datum/light_eater)
+	. = list()
+	SEND_SIGNAL(commisary, COMSIG_LIGHT_EATER_QUEUE, ., light_eater)
```
```suggestion
	SEND_SIGNAL(commisary, COMSIG_LIGHT_EATER_QUEUE, .)
```
This is already the source of the signal.

Rohesie commented on code/datums/elements/light_eater.dm:
```diff
@@ -0,0 +1,171 @@
+ * - [comissary][/atom]: The origin node of all of the light sources to search through.
+ * - [light_eater][/datum]: The light eater being applied to the target.
+ */
+/datum/element/light_eater/proc/table_buffet(atom/commisary, datum/light_eater)
```
What is `datum/light_eater`? The component? If so, it shouldn't be passed around.
This doesn't seem to be used, though.

TemporalOroboros replied:
It's not the component. I think I made it the parent to the light eater component in case whatever I'm hitting wants to react/change behavior according to that. It's a datum because I made the component compatible with reagents.

Rohesie replied:
> 
> 
> It's not the component. I think I made it the parent to the light eater component in case whatever I'm hitting wants to react/change behavior according to that. It's a datum because I made the component compatible with reagents.

Please rephrase, I didn't understand the answer.
What is `datum/light_eater` if not the component? Is this a reference to the element itself, the `src` of the proc? Or to something else?

TemporalOroboros replied:
It is the reference to the datum the light eater element is attached to.

Rohesie commented on code/datums/elements/light_eaten.dm:
```diff
@@ -0,0 +1,42 @@
+	if(eaten_light.light_power <= 0)
+		return NONE
+
+	eaten_light.light_power = 0
+	if(eaten_light.light_range > 0)
+		eaten_light.light_range = 0
+	if(eaten_light.light_on)
+		eaten_light.light_on = FALSE
```
Use the setter procs for these, (like `set_light_power()`), else it will break the light component.

TemporalOroboros replied:
Alright. I guess there's no harm in that.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Slight code tweaks

- Makes everything that changes light values go through the setters.
- Makes the light eater component compatible with the overlay lighting components.
- Clarifies some comments.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros commented:
I've done a bit of testing and it works.

Rohesie commented on code/modules/lighting/lighting_atom.dm:
```diff
@@ -112,42 +114,42 @@
 	if(new_power == light_power)
 		return
-	SEND_SIGNAL(src, COMSIG_ATOM_SET_LIGHT_POWER, new_power)
 	. = light_power
 	light_power = new_power
+	SEND_SIGNAL(src, COMSIG_ATOM_SET_LIGHT_POWER, new_power, .)
```
Why did you move the signal to afterwards?
We try to have them as soon as we can whenever possible, because that allows us to add the option to cancel it, for example, if needed.

TemporalOroboros replied:
Because this is the standard for setter procs with signals as far as I can tell.

Rohesie replied:
This is not the standard at all.

TemporalOroboros replied:
Done. More overhead, but done.

Rohesie commented on code/modules/vehicles/mecha/_mecha.dm:
```diff
@@ -1223,3 +1225,15 @@
+
+
+/// Special light eater handling
+/obj/vehicle/sealed/mecha/proc/on_light_eater(obj/vehicle/sealed/mecha, datum/source)
```
The mecha is the source of the signal, not the eater.
Can you fix these names?

TemporalOroboros replied:
Sure.

TemporalOroboros replied:
Done.

Rohesie commented on code/modules/power/lighting.dm:
```diff
@@ -843,6 +845,14 @@
 
+/obj/machinery/light/proc/on_light_eater(obj/machinery/light, datum/light_eater)
+	SIGNAL_HANDLER
+	if(status == 1)
```
What is `status == 1`? Could you fix this magic number?

TemporalOroboros replied:
Sure. I think I just copied this from the old proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -111,6 +112,7 @@
 		if(T && istype(radio) && istype(radio.keyslot))
 			radio.keyslot.forceMove(T)
 			radio.keyslot = null
+	UnregisterSignal(src, COMSIG_LIGHT_EATER_ACT)
```
Signals get automatically unregistered when the object is deleted already.

TemporalOroboros replied:
I know. But this is the standard format.

Rohesie replied:
Where is this standard taken from?

TemporalOroboros replied:
Done. More overhead, but done.

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+	Insert(user)
+
+/obj/item/organ/heart/nightmare/Insert(mob/living/carbon/M, special = 0)
+	..()
```
```suggestion
	. = ..()
```

TemporalOroboros replied:
Done

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+	if(blade && special != HEART_SPECIAL_SHADOWIFY)
+		M.visible_message("<span class='warning'>\The [blade] disintegrates!</span>")
+		QDEL_NULL(blade)
+	..()
```
```suggestion
	return ..()
```

TemporalOroboros replied:
Done

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+	user.temporarilyRemoveItemFromInventory(src, TRUE)
+	Insert(user)
+
+/obj/item/organ/heart/nightmare/Insert(mob/living/carbon/M, special = 0)
```
What does `special = 0` mean?

TemporalOroboros replied:
I'm pretty sure this just means that if it replaces a vital organ it kills by default. It was like that when I moved it.

TemporalOroboros replied:
I've replaced it with FALSE.

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+	desc = "An alien organ that twists and writhes when exposed to light."
+	icon = 'icons/obj/surgery.dmi'
+	icon_state = "demon_heart-on"
+	color = "#1C1C1C"
+	var/respawn_progress = 0
+	var/obj/item/light_eater/blade
+	decay_factor = 0
```
Can you put redefined variables over the newly-defined ones, and document new variables?

TemporalOroboros replied:
Sure.

TemporalOroboros replied:
Done.

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+/obj/item/organ/brain/nightmare/Remove(mob/living/carbon/M, special = 0)
+	if(shadowwalk)
+		M.RemoveSpell(shadowwalk)
+	..()
```
```suggestion
	return ..()
```

TemporalOroboros replied:
It was like that originally and I don't want to break something by changing it. I'll take a look.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -0,0 +1,101 @@
+	var/obj/effect/proc_holder/spell/targeted/shadowwalk/shadowwalk
+
+/obj/item/organ/brain/nightmare/Insert(mob/living/carbon/M, special = 0)
+	..()
```
```suggestion
	. = ..()
```

TemporalOroboros replied:
Same

TemporalOroboros replied:
Done

Rohesie commented on code/modules/antagonists/nightmare/nightmare.dm:
```diff
@@ -3,3 +3,10 @@
 	show_in_antagpanel = FALSE
 	show_name_in_check_antagonists = TRUE
 	show_to_ghosts = TRUE
+
+
+
+
+
+
+
```
What's this about?

TemporalOroboros replied:
I don't know, but I removed it.

Rohesie commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -1106,6 +1116,7 @@ GLOBAL_LIST_EMPTY(PDAs)
 
 /obj/item/pda/Destroy()
 	GLOB.PDAs -= src
+	UnregisterSignal(src, COMSIG_LIGHT_EATER_ACT)
```
No need to unregister source signals on destroy, that's done automatically already.

TemporalOroboros replied:
It's the standard for components and the like.

Rohesie replied:
Check the definition of `Destroy()` at the base.
This is already handled by default.

TemporalOroboros replied:
Done

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
A bunch of changes @Rohesie suggested.

- Adds a second set of signals to the light setters.
- Cleans up some organ code for nightmares
- Clarifies some signal args

TemporalOroboros commented:
I tested this and it works. I'm still waiting for ninja to get back to me on the unregistering signals bit.

TemporalOroboros committed:
Merge branch 'master' into light-eater

Rohesie commented:
Ping him on discord if you want an answer.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Sacrifices code clarity

- Removes explicit signal unregistration

TemporalOroboros commented:
Done and tested.

Rohesie commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -219,16 +217,32 @@
+///Called right before the atom changes the value of light_power to a different one, from base [atom/proc/set_light_power]: (new_power, old_power)
 #define COMSIG_ATOM_SET_LIGHT_POWER "atom_set_light_power"
-///Called right before the atom changes the value of light_color to a different one, from base atom/set_light_color(): (new_color)
+///Called right after the atom changes the value of light_power to a different one, from base of [/atom/proc/set_light_power]: (new_power, old_power)
```
Fix this please. Make the signal be sent before, not after.

TemporalOroboros replied:
I did. You told me to make two signals so I made two signals.

Rohesie replied:
Right, I see that now.
```suggestion
///Called right after the atom changes the value of light_power to a different one, from base of [/atom/proc/set_light_power]: (old_power)
```
No need to pass the new power as an argument, since the source is already being passed.

TemporalOroboros replied:
Done

Rohesie replied:
Update the comment as well, and apply this to all the new signals, not just this one.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
I don't know why this change was requested.

TemporalOroboros committed:
Simplifies signals and makes the light eater not runtime sometimes.

TemporalOroboros commented on code/datums/elements/light_eater.dm:
```diff
@@ -0,0 +1,171 @@
+			else if(isprojectile(source))
+				UnregisterSignal(source, COMSIG_PROJECTILE_ON_HIT)
+	else if(istype(source, /datum/reagent))
+		UnregisterSignal(source, COMSIG_REAGENT_EXPOSE_ATOM)
```
Done

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros commented:
Alright. I did it. Also tested.

TemporalOroboros committed:
Merge branch 'master' into light-eater

Rohesie reviewed (changes requested):
Merge conflict and one change request.
Should be ready for merge after.

TemporalOroboros committed:
Merge branch 'master' into light-eater

TemporalOroboros committed:
Whitespace fixes

TemporalOroboros commented:
Done and tested.

Rohesie commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -221,16 +219,32 @@
+#define COMSIG_ATOM_UPDATE_LIGHT_POWER "atom_update_light_power"
+///Called right before the atom changes the value of light_range to a different one, from base [atom/proc/set_light_range]: (new_range)
+#define COMSIG_ATOM_SET_LIGHT_RANGE "atom_set_light_range"
+///Called right after the atom changes the value of light_range to a different one, from base of [/atom/proc/set_light_range]: (new_range, old_range)
```
Sorry if I wasn't clear, but the same logic applies to all the new signals.
Send only the old value forward and update the comment.
The new value is already known since we're passing the source.

TemporalOroboros replied:
Done

TemporalOroboros committed:
Remove even more signal args for some reason

Rohesie commented on code/datums/components/overlay_lighting.dm:
```diff
@@ -107,12 +107,13 @@
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_COLOR, .proc/set_color)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_ON, .proc/on_toggle)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_FLAGS, .proc/on_light_flags_change)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_RANGE, .proc/set_range)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_POWER, .proc/set_power)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_COLOR, .proc/set_color)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_ON, .proc/on_toggle)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_FLAGS, .proc/on_light_flags_change)
```
Why are these changed? They worked fine with the pre-signal.
Revert this change, please.

TemporalOroboros replied:
Because they now go after the setter.
